File: src/lexer.c
#include "lexer.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// Utility function to initialize keyword lists
static const char** init_keywords(const char* keywords[], size_t count) {
    const char** kw = malloc(sizeof(char*) * count);
    for (size_t i = 0; i < count; i++) {
        kw[i] = keywords[i];
    }
    return kw;
}

// Create a new Lexer instance
Lexer* lexer_create(const char* source_code) {
    Lexer* lexer = malloc(sizeof(Lexer));
    lexer->source_code = strdup(source_code);
    lexer->tokens = NULL;
    lexer->token_count = 0;
    lexer->current_pos = 0;
    lexer->line_num = 1;

    // Initialize keyword lists
    static const char* keywords_array[] = {
        "SET", "DEFUN", "IF", "THEN", "ELSE", "ELSEIF", "ENDIF",
        "TIMES", "RETURN", "BREAK", "CONTINUE", "YIELD", "PASS"
    };
    lexer->keywords = init_keywords(keywords_array, sizeof(keywords_array) / sizeof(char*));
    lexer->keywords_count = sizeof(keywords_array) / sizeof(char*);

    static const char* loop_keywords_array[] = { "REPEAT", "WHILE" };
    lexer->loop_keywords = init_keywords(loop_keywords_array, sizeof(loop_keywords_array) / sizeof(char*));
    lexer->loop_keywords_count = sizeof(loop_keywords_array) / sizeof(char*);

    // Initialize other keyword lists similarly...
    // For brevity, not all lists are initialized here. Implement similarly as above.

    // Example for io_keywords
    static const char* io_keywords_array[] = { "PRINTLN", "PRINT", "INPUT", "OPEN", "WRITE", "RUN" };
    lexer->io_keywords = init_keywords(io_keywords_array, sizeof(io_keywords_array) / sizeof(char*));
    lexer->io_keywords_count = sizeof(io_keywords_array) / sizeof(char*);

    // Similarly initialize input_control_keywords, error_keywords, control_keywords, etc.

    return lexer;
}

// Destroy a Lexer instance
void lexer_destroy(Lexer* lexer) {
    if (!lexer) return;
    free(lexer->source_code);
    // Free tokens
    for (size_t i = 0; i < lexer->token_count; i++) {
        token_destroy(lexer->tokens[i]);
    }
    free(lexer->tokens);
    // Free keyword lists
    free((void*)lexer->keywords);
    free((void*)lexer->loop_keywords);
    free((void*)lexer->io_keywords);
    // Free other keyword lists similarly...
    free(lexer);
}

// Check if a word is a keyword in any list
int is_keyword(const Lexer* lexer, const char* word) {
    return is_keyword_in_list(lexer->keywords, lexer->keywords_count, word) ||
           is_keyword_in_list(lexer->loop_keywords, lexer->loop_keywords_count, word) ||
           is_keyword_in_list(lexer->io_keywords, lexer->io_keywords_count, word);
           // Add checks for other keyword lists...
}

// Check if a word is in a specific keyword list
int is_keyword_in_list(const char** list, size_t list_count, const char* word) {
    for (size_t i = 0; i < list_count; i++) {
        if (strcasecmp(list[i], word) == 0) {
            return 1;
        }
    }
    return 0;
}

// Process time with unit
void process_time_with_unit(const char* match_str, double* value, char* unit) {
    sscanf(match_str, "%lf%s", value, unit);
    if (strcmp(unit, "ms") == 0) {
        *value /= 1000.0;
    } else if (strcmp(unit, "m") == 0) {
        *value *= 60.0;
    } else if (strcmp(unit, "h") == 0) {
        *value *= 3600.0;
    }
    // 's' doesn't need conversion
}

// Tokenize the source code
Token** lexer_tokenize(Lexer* lexer, size_t* out_token_count) {
    size_t capacity = 128;
    lexer->tokens = malloc(sizeof(Token*) * capacity);
    lexer->token_count = 0;

    size_t length = strlen(lexer->source_code);
    char current_char;
    char buffer[256];
    size_t buffer_pos = 0;

    while (lexer->current_pos < length) {
        current_char = lexer->source_code[lexer->current_pos];

        // Handle whitespace
        if (isspace(current_char)) {
            if (current_char == '\n') {
                lexer->line_num++;
            }
            lexer->current_pos++;
            continue;
        }

        // Handle comments
        if (current_char == '#') {
            // Single-line comment
            if (lexer->source_code[lexer->current_pos + 1] != '*') {
                while (current_char != '\n' && lexer->current_pos < length) {
                    current_char = lexer->source_code[++lexer->current_pos];
                }
                continue;
            }
            // Multi-line comment (#* *#)
            lexer->current_pos += 2;
            while (!(lexer->source_code[lexer->current_pos] == '*' &&
                     lexer->source_code[lexer->current_pos + 1] == '#') &&
                   lexer->current_pos < length) {
                if (lexer->source_code[lexer->current_pos] == '\n') {
                    lexer->line_num++;
                }
                lexer->current_pos++;
            }
            lexer->current_pos += 2;
            continue;
        }

        // Handle identifiers and keywords
        if (isalpha(current_char) || current_char == '_') {
            buffer_pos = 0;
            while (isalnum(current_char) || current_char == '_') {
                buffer[buffer_pos++] = current_char;
                lexer->current_pos++;
                current_char = lexer->source_code[lexer->current_pos];
            }
            buffer[buffer_pos] = '\0';
            if (is_keyword(lexer, buffer)) {
                Token* token = token_create("KEYWORD", buffer, lexer->line_num, NULL, NULL);
                // Add token to list
                if (lexer->token_count >= capacity) {
                    capacity *= 2;
                    lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
                }
                lexer->tokens[lexer->token_count++] = token;
            } else {
                Token* token = token_create("ID", buffer, lexer->line_num, NULL, NULL);
                // Add token to list
                if (lexer->token_count >= capacity) {
                    capacity *= 2;
                    lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
                }
                lexer->tokens[lexer->token_count++] = token;
            }
            continue;
        }

        // Handle numbers (integers and floats)
        if (isdigit(current_char)) {
            buffer_pos = 0;
            int is_float = 0;
            while (isdigit(current_char) || current_char == '.') {
                if (current_char == '.') {
                    is_float = 1;
                }
                buffer[buffer_pos++] = current_char;
                lexer->current_pos++;
                current_char = lexer->source_code[lexer->current_pos];
            }
            buffer[buffer_pos] = '\0';
            if (is_float) {
                Token* token = token_create("FLOAT", buffer, lexer->line_num, NULL, NULL);
                if (lexer->token_count >= capacity) {
                    capacity *= 2;
                    lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
                }
                lexer->tokens[lexer->token_count++] = token;
            } else {
                Token* token = token_create("INTEGER", buffer, lexer->line_num, NULL, NULL);
                if (lexer->token_count >= capacity) {
                    capacity *= 2;
                    lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
                }
                lexer->tokens[lexer->token_count++] = token;
            }
            continue;
        }

        // Handle string literals
        if (current_char == '"') {
            lexer->current_pos++; // Skip the opening quote
            buffer_pos = 0;
            current_char = lexer->source_code[lexer->current_pos];
            while (current_char != '"' && lexer->current_pos < length) {
                buffer[buffer_pos++] = current_char;
                lexer->current_pos++;
                current_char = lexer->source_code[lexer->current_pos];
            }
            buffer[buffer_pos] = '\0';
            lexer->current_pos++; // Skip the closing quote
            Token* token = token_create("STRING", buffer, lexer->line_num, NULL, NULL);
            if (lexer->token_count >= capacity) {
                capacity *= 2;
                lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
            }
            lexer->tokens[lexer->token_count++] = token;
            continue;
        }

        // Handle operators and punctuations
        // Example: +, -, *, /, ==, !=, etc.
        // This section can be expanded based on the operators defined in the Python lexer

        // For simplicity, handle single-character operators
        buffer[0] = current_char;
        buffer[1] = '\0';
        Token* token = token_create("OPERATOR", buffer, lexer->line_num, NULL, NULL);
        lexer->current_pos++;
        if (lexer->token_count >= capacity) {
            capacity *= 2;
            lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
        }
        lexer->tokens[lexer->token_count++] = token;
    }

    // Append EOF token
    Token* eof_token = token_create("EOF", NULL, lexer->line_num, NULL, NULL);
    if (lexer->token_count >= capacity) {
        capacity += 1;
        lexer->tokens = realloc(lexer->tokens, sizeof(Token*) * capacity);
    }
    lexer->tokens[lexer->token_count++] = eof_token;

    // Connect tokens
    lexer->tokens = connect_tokens(lexer->tokens, lexer->token_count);

    *out_token_count = lexer->token_count;
    return lexer->tokens;
}

// Connect tokens for bidirectional traversal
Token** connect_tokens(Token** tokens, size_t token_count) {
    for (size_t i = 0; i < token_count; i++) {
        if (i > 0) {
            tokens[i]->previous_token = tokens[i - 1];
        }
        if (i < token_count - 1) {
            tokens[i]->next_token = tokens[i + 1];
        }
    }
    return tokens;
}

File: python/lexer.py
import re
import logging
from .errors import InvalidNumberError, SyntaxError
from .ast_nodes import Token

# Configure logger for this module
logger = logging.getLogger(__name__)

class Lexer:
    def __init__(self, source_code):
        # Core language keywords
        self.keywords = {
            "SET",
            "DEFUN",
            "LAMBDA",
            "IF",
            "THEN", 
            "ELSE",
            "ELSEIF",
            "ENDIF",
            "TIMES",
            "RETURN",
            "BREAK", 
            "CONTINUE",
            "YIELD",
            "PASS",
            "POINT"
        }

        self.loop_keywords = {
            "REPEAT",
            "WHILE",
        }

        # I/O and interaction keywords
        self.io_keywords = {
            "PRINTLN",
            "PRINT", 
            "INPUT",
            "OPEN",
            "WRITE",
            "RUN",
        }

        # Mouse and keyboard control keywords
        self.input_control_keywords = {
            "MOVE",
            "MOUSE",
            "WAIT",
            "RELEASE",
            "HOLD",
            "PRESS",
            "FOCUS",
            "SCROLL",
        }

        # Error handling keywords
        self.error_keywords = {
            "TRY",
            "EXCEPT", 
            "FINALLY",
            "ERROR",
        }

        # Control flow keywords
        self.control_keywords = {
            "EXIT",
            "CONTINUE", 
            "BREAK",
            "RAISE",
        }

        # Generator keywords
        self.generator_keywords = {
            "YIELD",
            "FROM",
        }

        # Type keywords
        self.type_keywords = {
            "APP",
            "WINDOW",
            "KEY",
            "BUTTON",
        }

        # Target/direction keywords
        self.target_keywords = {
            "TO",
        }

        # Assertion/comparison keywords
        self.assertion_keywords = {
            "EXISTS",
            "IS",
            "IN",
            "AT",
        }

        # Keyboard key definitions
        self.keyboard_keys = {
            "FN",
            "BACKSPACE",
            "ENTER", 
            "SPACE",
            "TAB",
            "(?:L|R|)CTRL",
            "(?:L|R|)ALT",
            "(?:L|R|)SHIFT",
            "(?:L|R|)WIN",
            "DEL",
            "DELETE",
            "END",
            "HOME",
            "INSERT",
            "PG_(?:U|DOWN)",
            "ARROW_(?:LEFT|RIGHT|UP|DOWN)",
            "ESC",
            "CAPS_LOCK",
            "[a-zA-Z]",
            "F(?:1[0-2]|[1-9])",
        }

        # Mouse button definitions
        self.mouse_keys = {
            "LEFT",
            "RIGHT", 
            "MIDDLE",
            "WHEEL_UP",
            "WHEEL_DOWN",
            "SCROLL_UP",  # how many times to scroll up (extension of WHEEL_UP)
            "SCROLL_DOWN",  # how many times to scroll down (extension of WHEEL_DOWN)
        }

        # Boolean literals
        self.boolean_values = {
            "TRUE",
            "FALSE"
        }

        self.source_code = source_code
        self.tokens = []
        self.current_pos = 0  # Initialize current_pos
        self.line_num = 1  # Initialize line number

        # Define token patterns as a dictionary for more efficient lookup
        self.token_patterns = {
            # Comments - both single line (#) and multi-line (#* *#)
            "COMMENT": r"#\*[\s\S]*?\*#|#.*",

            # Keywords (only pure uppercase or lowercase)
            "TYPE_KEYWORD": r"\b(?:" + "|".join([f"({k}|{k.lower()})" for k in self.type_keywords]) + r")\b",
            "KEYWORD": r"\b(?:" + "|".join([f"({k}|{k.lower()})" for k in (self.keywords | self.io_keywords | self.input_control_keywords | self.error_keywords)]) + r")\b",
            "KEYWORD_ASSERTION": r"\b(?:" + "|".join([f"({k}|{k.lower()})" for k in self.assertion_keywords]) + r")\b",
            "KEYWORD_TARGET": r"\b(?:" + "|".join([f"({k}|{k.lower()})" for k in self.target_keywords]) + r")\b",
            
            # Special keys (case-sensitive)
            "KEYBOARD_KEY": r"\b(?:" + "|".join(self.keyboard_keys) + r")\b",
            "MOUSE_KEY": r"\b(?:" + "|".join(self.mouse_keys) + r")\b",

            # Loop keywords
            "LOOP": r"\b(?:" + "|".join([f"({k}|{k.lower()})" for k in self.loop_keywords]) + r")\b",
            
            # Literals
            "FLOAT": r"\b\d+\.\d+\b",
            "INT": r"\b\d+\b",
            "TIME": r"\b\d+(?:\.\d+)?[smh]\b|\b\d+ms\b", 
            "STR": r'(["\'])(?:\\.|[^\\\1])*?\1',
            "BOOL": r"\b(?:TRUE|FALSE|true|false)\b",

            # Increment/decrement operators
            "INCREMENT": r"\+\+",
            "DECREMENT": r"--",

            # Identifiers
            "ID": r"\b[a-zA-Z_][a-zA-Z0-9_]*\b",

            # Operators (order matters - longer patterns first)
            "COMP_OP": r"===|!==|==|!=|<=|>=|<|>|\|>",
            "OP_ASSIGN": r"\+=|-=|\*=|/=|%=|&=|\^=|<<=|>>=|=",
            "BITWISE_OP": r"\||&|\^|~|<<|>>",
            "OP": r"\*\*|//|[+\-*/]",  # Added ** and // as single operators
            "TERMINATOR": r";",

            # Parentheses
            "L_PAREN": r"\(",
            "R_PAREN": r"\)",

            # Braces
            "L_BRACE": r"\{",
            "R_BRACE": r"\}",

            # Commas
            "COMMA": r",",

            # Type Hint Token
            "TYPE_HINT": r":",

            # Newlines and whitespace
            "NEWLINE": r"\n",
            "SKIP": r"[ \t]+",
            "MISMATCH": r"."
        }

        # Update the time unit pattern in token_patterns
        self.token_patterns['TIME'] = r'(\d+(?:\.\d+)?)(ms|s|m|h)'
        
        # Compile the regex pattern with named groups
        self.token_regex = "|".join(f"(?P<{kind}>{pattern})" for kind, pattern in self.token_patterns.items())

        logger.debug("Lexer initialized with source code length %d.", len(source_code))

    def tokenize(self):
        logger.info("Starting tokenization process.")
        previous_token = None

        for match in re.finditer(self.token_regex, self.source_code):
            kind = match.lastgroup
            value = match.group() 

            # Start of Selection
            logger.debug(
                "Matched token: %s with value '%s' at line %d.",
                kind,
                value.encode('unicode_escape').decode(),
                self.line_num
            )

            if kind == "NEWLINE":
                self.line_num += 1
                logger.debug("Encountered NEWLINE. Incremented line number to %d.", self.line_num)
                continue
            elif kind in ["COMMENT", "SKIP"]:
                logger.debug("Skipping token of kind '%s'.", kind)
                continue

            if kind == "FLOAT":
                try:
                    value = float(value)
                except ValueError:
                    logger.error("Invalid FLOAT value '%s' at line %d.", value, self.line_num)
                    raise InvalidNumberError(f"Invalid FLOAT value '{value}' at line {self.line_num}")
            elif kind == "INT":
                value = int(value)
            elif kind == "TIME":
                value = self.process_time_with_unit(value)
                kind = "TIME"
            elif kind == "BOOL":
                value = value.upper() == "TRUE"  # Convert to Python boolean, case-insensitive
            elif kind == "MISMATCH":
                error_msg = f"Unexpected character at line {self.line_num}: {value}"
                logger.error(error_msg)
                raise SyntaxError(error_msg)

            # Convert keyword values to uppercase for consistency
            if kind in ["KEYWORD", "TYPE_KEYWORD", "KEYWORD_ASSERTION", "KEYWORD_TARGET", "LOOP"]:
                value = value.upper()

            # Handle KEYBOARD_KEY based on previous token
            if kind == "KEYBOARD_KEY":
                if previous_token and previous_token.kind == "TYPE_KEYWORD" and previous_token.value == "KEY":
                    logger.debug("Valid KEYBOARD_KEY '%s' found.", value)
                    pass  # Valid KEYBOARD_KEY
                else:
                    logger.debug("Invalid KEYBOARD_KEY context for '%s'. Treating as ID.", value)
                    kind = "ID"  # Treat as ID if not valid in context

            # Create Token with line number
            current_token = Token(kind, value, self.line_num, previous_token)
            if previous_token:
                previous_token.next_token = current_token
            self.tokens.append(current_token)
            logger.debug("Appended token: %s, line: %d", current_token, self.line_num)
            previous_token = current_token

        # Add EOF token with current line number
        eof_token = Token("EOF", None, self.line_num, previous_token)
        if previous_token:
            previous_token.next_token = eof_token
        self.tokens.append(eof_token)
        logger.debug("Appended EOF token at line %d.", self.line_num)

        logger.info("Tokenization process completed successfully.")
        return self.connect_tokens()

    def process_time_with_unit(self, match_str):
        logger.debug("Processing TIME token '%s'.", match_str)
        time_match = re.match(r'(\d+(?:\.\d+)?)(ms|s|m|h)', match_str)
        value = float(time_match.group(1))
        unit = time_match.group(2)

        if unit == 'ms':
            value = value / 1000  # milliseconds to seconds
        elif unit == 'm':
            value = value * 60    # minutes to seconds
        elif unit == 'h':
            value = value * 3600  # hours to seconds
        # 's' doesn't need conversion

        logger.debug("Converted TIME token '%s' to (%f, '%s').", match_str, value, 's')
        return (value, 's')

    def connect_tokens(self):
        # This method ensures that each token has correct previous and next references
        logger.debug("Connecting tokens for bidirectional traversal.")
        for i in range(len(self.tokens)):
            if i > 0:
                self.tokens[i].previous_token = self.tokens[i - 1]
            if i < len(self.tokens) - 1:
                self.tokens[i].next_token = self.tokens[i + 1]
        logger.debug("Tokens connected successfully.")
        return self.tokens

    def parse_numbers(self, number_str):
        # Check if the number is an integer
        try:
            int_value = int(number_str)
            logger.debug("Parsed integer number: %d", int_value)
            return int_value
        except ValueError:
            pass

        # Check if the number is a float
        try:
            float_value = float(number_str)
            logger.debug("Parsed float number: %f", float_value)
            return float_value
        except ValueError:
            pass

        error_msg = f"'{number_str}' is not a valid number."
        logger.error(error_msg)
        raise InvalidNumberError(error_msg)

    def get_next_token(self):
        if not self.tokens:
            logger.debug("Token list empty. Starting tokenization.")
            self.tokenize()

        if self.current_pos < len(self.tokens):
            token = self.tokens[self.current_pos]
            self.current_pos += 1
            logger.debug("Retrieved next token: %s", token)
            return token
        else:
            logger.debug("No more tokens. Returning EOF.")
            return Token("EOF", None, self.line_num)